### PR TITLE
249 filters query update or to xor in exclusive categories

### LINKED ISSUE #249: Filters query update OR to XOR in exclusive categories
Link to Thread
https://gnardog.slack.com/archives/C03KGHSSTNC/p1673621197886439

### PR DESCRIPTION
**Pull request by Elkrival**
The query was providing a result of OR, this updates the query to return a combination of AND with all categories then having an XOR. Due to the current data structure the cursor returns a grouping of the data by the filter category and we verify if any of documents match the category and requested filter.

https://gnardog.slack.com/archives/C03KGHSSTNC/p1673621197886439

### CODE REVIEW

File: server/controllers/chartsController/index.js [hunk 1]
```diff
@@ -116,11 +117,17 @@ const getAllSubjects = async ({ dataDb, chart, userAccess, filters }) => {
...
-      const data = await dataDb
-        .collection(doc.collection)
-        .findOne({ $or: mongoFilterQuery })
+      const data = await dataDb.collection(doc.collection).aggregate(aggregate)
+      const subjectCriteriaData = await data.next()
+      const isSubjectInFilteredQuery = activeFilters
+        .map(
+          (requestedFilter) => subjectCriteriaData[requestedFilter].length > 0
+        )
+        .every((value) => value === true)
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js]:
```suggestion
        .every(Boolean)
```

File: server/controllers/chartsController/index.js [hunk 2]
```diff
@@ -116,11 +117,17 @@ const getAllSubjects = async ({ dataDb, chart, userAccess, filters }) => {
     )
 
     for await (let doc of criteriaCursor) {
-      const data = await dataDb
-        .collection(doc.collection)
-        .findOne({ $or: mongoFilterQuery })
+      const data = await dataDb.collection(doc.collection).aggregate(aggregate)
+      const subjectCriteriaData = await data.next()
```
Comment (mikestone14) [inline server/controllers/chartsController/index.js:123]: what does `data.next()` do?

**Review by mikestone14 — COMMENTED** [review]
This looks good to me but I don't have a ton of Mongo experience so probably worth having someone else on the team take a look before merging this in.

File: server/controllers/chartsController/index.js (hunk 2 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js:123]: The aggregate method returns a cursor, the cursor is a pointer and to access the data streaming in the cursor we can use different methods like toArray() etc, In the aggregate I'm returning one document with an array of data and to access that type of cursor response we use the next() method

File: server/controllers/chartsController/index.js (hunk 1 above)
Comment (Elkrival) [reply server/controllers/chartsController/index.js]: https://github.com/AMP-SCZ/dpdash/pull/250/commits/e5140b25a145e5811e26abcd2d4638ed770f1e75

**Review by mikestone14 — APPROVED** [review]

**Pull request merged**